### PR TITLE
Track and show folder last scan time (fixes #3143)

### LINKED ISSUE #3143: Date and time of last scan
Sometimes it would be helpful to display date/time of last scan of each folder, especially if you have lots of data and long scan intervals

### PR DESCRIPTION
**Pull request by majedev**
### Purpose

Keep track of folders last scan time and show it on GUI.

### Testing

Tested with 2 folders using manual scan (by pressing scan button on GUI) and auto rescan interval.

### Screenshots

screenshot

### Documentation

REST API documentation for `/rest/stats/folder` endpoint should be updated to include `lastScan` in JSON reply

### CODE REVIEW

File: lib/stats/folder.go [hunk 1]
```diff
@@ -59,8 +60,21 @@ func (s *FolderStatisticsReference) ReceivedFile(file string, deleted bool) {
 	s.ns.PutBool("lastFileDeleted", deleted)
 }
 
+func (s *FolderStatisticsReference) ScanCompleted() {
+	s.ns.PutTime("lasScan", time.Now())
+}
+
+func (s *FolderStatisticsReference) GetLastScanTime() time.Time {
+	lasScan, ok := s.ns.Time("lasScan")
```
Comment (AudriusButkevicius) [inline lib/stats/folder.go]: Typo

File: gui/default/index.html [hunk 2]
```diff
@@ -360,6 +360,14 @@ <h4 class="panel-title">
                       <th><span class="fa fa-fw fa-share-alt"></span>&nbsp;<span translate>Shared With</span></th>
                       <td class="text-right">{{sharesFolder(folder)}}</td>
                     </tr>
+                    <tr ng-if="folder.type != 'readonly'">
```
Comment (AudriusButkevicius) [inline gui/default/index.html]: Readonly are still scanned

File: gui/default/index.html [hunk 3]
```diff
@@ -360,6 +360,14 @@ <h4 class="panel-title">
...
+                      <th><span class="fa fa-fw fa-clock-o"></span>&nbsp;<span translate>Last Scan Time</span></th>
+                      <td class="text-right">
+                        <span>
+                          {{folderStats[folder.id].lastScan | date:'yyyy-MM-dd HH:mm:ss'}}
```
Comment (AudriusButkevicius) [inline gui/default/index.html]: Probably needs a special case if its 0, see lastSeen for devices.

File: lib/model/model.go [hunk 4]
```diff
@@ -1602,6 +1602,7 @@ func (m *Model) internalScanFolderSubdirs(folder string, subs []string) error {
 		m.updateLocalsFromScanning(folder, batch)
 	}
 
+	m.folderStatRef(folder).ScanCompleted()
```
Comment (AudriusButkevicius) [inline lib/model/model.go]: So I wonder if we should update this if subs is not nil (for example inotify asking us to scan a specific file). Open to suggestions/discussions.

@uok @Zillode @canton7 @Nutomic @calmh

File: lib/stats/folder.go [hunk 5]
```diff
@@ -59,8 +60,21 @@ func (s *FolderStatisticsReference) ReceivedFile(file string, deleted bool) {
 	s.ns.PutBool("lastFileDeleted", deleted)
 }
 
+func (s *FolderStatisticsReference) ScanCompleted() {
+	s.ns.PutTime("lasScan", time.Now())
```
Comment (AudriusButkevicius) [inline lib/stats/folder.go]: Typo

**Comment by canton7** [comment]
> So I wonder if we should update this if subs is not nil (for example inotify asking us to scan a specific file)

So that "last scan" is updated when something pokes the REST API asking for a scan of a particular file/dir? I'd say so: it's a good indication that Syncthing is keeping on top of changes you're making to files.

**Comment by calmh** [comment]
Yeah any kind of scanning activity works for me

**Comment by AudriusButkevicius** [comment]
So apart for the other two comments (regarding readonly and handling 0 time) I am happy to merge.

**Comment by uok** [comment]
I'd prefer "last scan" which is shorter and easier to translate. Other than that thanks for implementing :smile:

**Comment by AudriusButkevicius** [comment]
Or `Last Scan(ned)? At`, because now it somewhat implies duration.

**Comment by uok** [comment]
Grammatically you are maybe right, I simply took "last update: date/time" from AV scanners as a reference

**Comment by AudriusButkevicius** [comment]
Ok let's just call it 'Last Scan' and get this merged.

**Comment by calmh** [comment]
@majedev Do a `go run build.go translate` as well to update lang-en.json

**Comment by calmh** [comment]
@st-review lgtm

**Comment by st-review** [comment]
@calmh: Noted! Need another LGTM or explicit merge command.

**Comment by calmh** [comment]
(New feature being tested with the bot; it'll note "lgtm"s on the commit when merging, and merge if it gets two of them from different people.)

**Comment by AudriusButkevicius** [comment]
@st-review lgtm

**Comment by st-review** [comment]
:ok_hand: Merged as 48245effdfe2a9c656983b82e4f8d0c8a1220965. Thanks, @majedev!

**Pull request closed without merging**